sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/sqs/SqsUnboundedReaderTest.java

```diff
@@ -17,113 +17,155 @@
-                .withSqsClientProvider(SqsClientProviderMock.of(client))
-                .withMaxNumRecords(messages.size()));
+
+    when(mockSource.getRead().sqsClientProvider()).thenReturn(SqsClientProviderMock.of(client));
+    when(mockSource.getRead().queueUrl()).thenReturn(queueUrl);
   }
 
   @Test
   public void testReadOneMessage() throws IOException {
-    setupOneMessage();
-    UnboundedSource.UnboundedReader<SqsMessage> reader =
-        source.createReader(pipeline.getOptions(), null);
+    setupMessages(DATA);
+    SqsUnboundedReader reader = new SqsUnboundedReader(mockSource, null);
+
     // Read one message.
     assertTrue(reader.start());
     assertEquals(DATA, reader.getCurrent().getBody());
     assertFalse(reader.advance());
+
     // ACK the message.
     UnboundedSource.CheckpointMark checkpoint = reader.getCheckpointMark();
     checkpoint.finalizeCheckpoint();
     reader.close();
   }
 
   @Test
-  public void testTimeoutAckAndRereadOneMessage() throws IOException {
-    setupOneMessage();
-    UnboundedSource.UnboundedReader<SqsMessage> reader =
-        source.createReader(pipeline.getOptions(), null);
-    SqsClient sqsClient = embeddedSqsRestServer.getClient();
+  public void testAckDeletedMessage() throws IOException {
+    setupMessages(DATA);
+    SqsUnboundedReader reader = new SqsUnboundedReader(mockSource, null);
+
+    // Read one message
+    assertTrue(reader.start());
+    assertEquals(DATA, reader.getCurrent().getBody());
+    String receiptHandle = reader.getCurrent().getReceiptHandle();
+    assertFalse(reader.advance());
+
+    // Simulate already ACKed message after re-delivery to different reader
+    testCase
+        .getClient()
+        .deleteMessage(b -> b.queueUrl(testCase.getQueueUrl()).receiptHandle(receiptHandle));
+
+    // Now ACK the message.
+    UnboundedSource.CheckpointMark checkpoint = reader.getCheckpointMark();
+    checkpoint.finalizeCheckpoint();
+    reader.close();
+  }
+
+  @Test
+  @Ignore("Behavior of SQSRestServer is broken")
+  public void testExtendDeletedMessage() throws IOException {
+    setupMessages(DATA);
+    Clock clock = mock(Clock.class);
+    when(clock.millis()).thenReturn(currentTimeMillis());
+
+    SqsUnboundedReader reader = new SqsUnboundedReader(mockSource, null, clock);
+
+    // Read one message
+    assertTrue(reader.start());
+    assertEquals(DATA, reader.getCurrent().getBody());
+
+    // Simulate already ACKed message after re-delivery to different reader
+    String receiptHandle = reader.getCurrent().getReceiptHandle();
+    testCase
+        .getClient()
+        .deleteMessage(b -> b.queueUrl(testCase.getQueueUrl()).receiptHandle(receiptHandle));
+
+    // Forward time to force extension of visibility
+    when(clock.millis()).thenReturn(currentTimeMillis() + reader.getVisibilityTimeoutMs() * 8 / 10);
+
+    // Advancing the reader will attempt extending the visibility
+    assertFalse(reader.advance());
```

Comment:
should be false because `reader.start()` has read the first message and it should not have been resent because invalidated ? Right ? If so, please update the comment to explain more

Reply:
👍